tt_metal/impl/allocator/algorithms/free_list.cpp

```diff
@@ -424,6 +424,64 @@ void FreeList::dump_blocks(std::ofstream &out) const {
     out << "\n";
 }
 
+void FreeList::shrink_size(DeviceAddr shrink_size, bool bottom_up) {
```

Comment:
how does this work for DRAM where we allocate bottom_up? we would have to have to batch up DRAM addresses or is this only supported for L1?

Reply:
The current change needed for sub device is just to support splitting L1, we aren't planning to split DRAM at this time so I've only added support for the case we need for sub device.
Regular DRAM buffers are allocated bottom up, but the program binaries are allocated top down to avoid fragmentation, so DRAM would require more work to have a generic solution.

Reply:
should there be asserts in the allocator::shrink_size then?

Reply:
allocator::shrink_size asserts that bottom_up is true and that the lowest occupied address is greater than the shrink size. This should be sufficient and we shouldn't need to explicitly assert from dram/l1. This functionally works for dram as well, the issue is just how we decide to allocate dram at the top level isn't conducive to being able to use this fn in its current state.

Reply:
So post a DRAM allocation, we'll hit this assert: `Shrinking to a size that would cut into allocated memory is not supported` right?

Reply:
Yes

Reply:
If this is a feature with limited exposure, i.e. only used by the SubDevice infra, this may be okay. If for some reason we think this can become user facing code, it might make sense to just assert out for DRAM given that its a constrained use-case.

Reply:
Currently we don't expose the device allocator directly to the user so it shouldn't be user facing code. If user is creating/managing their own allocators where they could interface directly with this then they'd probably be in control of which directions buffers can be allocated, so they aren't subject to the same issues we are with how our infra allocates buffers. Doesn't seem like it requires a hard assertion from us to not be DRAM.

Reply:
Seems reasonable to me, thanks!